extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java

```diff
@@ -48,4 +87,78 @@ public String getImplementationName() {
     public ParquetInstructions readInstructions() {
         return readInstructions;
     }
+
+    /**
+     * Precedence-wise this implementation compares {@code order}, then {@code dataSequenceNumber}, then
+     * {@code fileSequenceNumber}, then {@code manifestSequenceNumber}, then {@code dataFilePos}, and finally compares
+     * {@code uri}.
+     *
+     * @inheritDoc
+     */
+    @Override
+    public int compareTo(@NotNull final TableLocationKey other) {
+        if (other instanceof IcebergTableParquetLocationKey) {
+            // TODO(deephaven-core#5989): Add unit tests for the ordering of data files
+            final IcebergTableParquetLocationKey otherTyped = (IcebergTableParquetLocationKey) other;
+            int comparisonResult;
+            if ((comparisonResult = Integer.compare(order, otherTyped.order)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(dataSequenceNumber, otherTyped.dataSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(fileSequenceNumber, otherTyped.fileSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(manifestSequenceNumber, otherTyped.manifestSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(dataFilePos, otherTyped.dataFilePos)) != 0) {
+                return comparisonResult;
+            }
+            return uri.compareTo(otherTyped.uri);
+        }
+        // When comparing with non-iceberg location key, we want to compare both partitions and URI
+        return super.compareTo(other);
```

Comment:
This is a good first step, but I'm realizing we have some trouble here.
We may have neglected this, but TLKs really must be comparable to all TLKs, not just TLKs of the same provenance.
We need fallback comparisons.

We have two paths forward:
1. Keep behavior as-is, and defer to a new issue:  Go back to throwing an exception on this line for now. Create an issue to define an ordering.
2. Fix the problem now.

My proposal for fixing, whether we do it here or in a new issue:
1. Make unrelated implementations compare to one another based on fully-qualified class name. This imposes an arbitrary but consistent ordering, and should roughly keep distinct sources together.
2. Make PartitionedTLK use order + partitions, followed by classname for tie-breaking.
3. Make URITLK order + partitions, then break ties by URI (if the other is a URITLK), and lastly break ties by classname.
4. Make IcebergTLK compare to other IcbergTLKs by Catalog name and TableIdentifier (which should also be added to equals and hashcode), then as currently defined in this PR, then to other URITLKs as defined by the URITLK rules, and so on.

Might want opinions from @abaranec and @devinrsmith